key-rotator/storage/manifest.go

```diff
@@ -0,0 +1,265 @@
+	}
+	return nil
+}
+
+// GetDataShareProcessorSpecificManifest gets the specific manifest for the
+// specified data share processor and returns it, if it exists and is
+// well-formed. If the manifest does not exist, an error wrapping
+// ErrObjectNotExist will be returned.
+func (m Manifest) GetDataShareProcessorSpecificManifest(ctx context.Context, dataShareProcessorName string) (manifest.DataShareProcessorSpecificManifest, error) {
+	key := m.keyFor(dataShareProcessorName)
+	manifestBytes, err := m.ds.get(ctx, key)
+	if err != nil {
+		return manifest.DataShareProcessorSpecificManifest{}, fmt.Errorf("couldn't get manifest from %q: %w", key, err)
+	}
+	var dspsm manifest.DataShareProcessorSpecificManifest
+	if err := json.Unmarshal(manifestBytes, &dspsm); err != nil {
+		return manifest.DataShareProcessorSpecificManifest{}, fmt.Errorf("couldn't unmarshal manifest from JSON: %w", err)
+	}
+	return dspsm, nil
+}
+
+// GetIngestorGlobalManifest gets the ingestor global manifest, if it exists
+// and is well-formed. If the manifest does not exist, an error wrapping
+// ErrObjectNotExist will be returned.
+func (m Manifest) GetIngestorGlobalManifest(ctx context.Context) (manifest.IngestorGlobalManifest, error) {
+	key := m.keyFor(ingestorGlobalManifestDataShareProcessorName)
+	manifestBytes, err := m.ds.get(ctx, key)
+	if err != nil {
+		return manifest.IngestorGlobalManifest{}, fmt.Errorf("couldn't get manifest from %q: %w", key, err)
+	}
+	var igm manifest.IngestorGlobalManifest
+	if err := json.Unmarshal(manifestBytes, &igm); err != nil {
+		return manifest.IngestorGlobalManifest{}, fmt.Errorf("couldn't unmarshal manifest from JSON: %w", err)
+	}
+	return igm, nil
+}
+
+func (m Manifest) keyFor(dataShareProcessorName string) string {
+	return path.Join(m.keyPrefix, fmt.Sprintf("%s-manifest.json", dataShareProcessorName))
+}
+
+// datastore represents a given key/value object store backing a Manifest. It
+// includes functionality for getting & putting individual objects by key,
+// specialized for small objects (i.e. no streaming support).
+type datastore interface {
+	// get gets the content of a given key, or returns an error if it can't.
+	// If the key does not exist, an error wrapping ErrObjectNotExist is
+	// returned.
+	get(ctx context.Context, key string) ([]byte, error)
+
+	// put puts the given content to the given key, or returns an error if it
+	// can't.
+	put(ctx context.Context, key string, data []byte) error
+}
+
+type gcsDatastore struct {
+	gcs    *storage.Client
+	bucket string
+}
+
+var _ datastore = gcsDatastore{} // verify gcsDatastore satsifies datastore.
+
+func (ds gcsDatastore) get(ctx context.Context, key string) ([]byte, error) {
+	r, err := ds.gcs.Bucket(ds.bucket).Object(key).NewReader(ctx)
+	if err != nil {
+		if err == storage.ErrObjectNotExist {
+			err = ErrObjectNotExist
+		}
+		return nil, fmt.Errorf("couldn't retrieve gs://%s/%s: %w", ds.bucket, key, err)
+	}
+	defer r.Close() // errchecked Close is below
```

Comment:
Ooh, nice, this is much better than checking whether `Close()` failed on every error path. I'm a little concerned about the behavior of calling `Close()` twice, though:

> [The behavior of Close after the first call is undefined. Specific implementations may document their own behavior.](https://pkg.go.dev/io#Closer)

[The docs on `storage.Reader.Close()`](https://pkg.go.dev/cloud.google.com/go/storage#Reader.Close) don't say anything about whether it's safe. WDYT?

Orthogonally: I was pedantic about this in the previous `manifest-updater/manifest/storage.go`, but I think we only need to be super careful about checking `Close()` on the write path. On the read path, we will know that our reads failed because we can't parse a JSON object.

Reply:
You're right -- I'm pretty sure this is safe in practice, but it's not promised to be safe by any of the APIs.

I switched to a defer, which will always call Close exactly once.